tests_e2e/scenarios/lib/vm_extension.py

```diff
@@ -0,0 +1,239 @@
+            if expected_version is not None:
+                # Compare only the major and minor versions (i.e. the first 2 items in the result of split())
+                installed_version = instance_view.type_handler_version
+                assert_that(expected_version.split(".")[0:2]).described_as("Unexpected extension version").is_equal_to(installed_version.split(".")[0:2])
+
+            assert_that(instance_view.statuses).described_as(f"Expected 1 status, got: {instance_view.statuses}").is_length(1)
+            status = instance_view.statuses[0]
+
+            if expected_message is not None:
+                assert_that(expected_message in status.message).described_as(f"{expected_message} should be in the InstanceView message ({status.message})").is_true()
+
+            assert_that(status.code).described_as("InstanceView status code").is_equal_to(expected_status_code)
+
+            if assert_function is not None:
+                assert_function(instance_view)
+
+        log.info("The instance view matches the expected values")
+
+    @abstractmethod
+    def delete(self) -> None:
+        """
+        Performs a delete operation on the extension
+        """
+
+    @property
+    @abstractmethod
+    def _ExtensionType(self) -> Type:
+        """
+        Type of the extension object for the virtual machine or scale set (i.e. VirtualMachineExtension or VirtualMachineScaleSetExtension)
+        """
+
+    @property
+    @abstractmethod
+    def _begin_create_or_update(self) -> Callable[[str, str, str, Any], LROPoller[Any]]:  # "Any" can be VirtualMachineExtension or VirtualMachineScaleSetExtension
+        """
+        The begin_create_or_update method for the virtual machine or scale set extension
+        """
+
+    @property
+    @abstractmethod
+    def _get(self) -> Any:  # VirtualMachineExtension or VirtualMachineScaleSetExtension
+        """
+        The get method for the virtual machine or scale set extension
+        """
+
+    def __str__(self):
+        return f"{self._identifier}"
+
+
+class VmExtension(_VmExtensionBaseClass):
+    """
+    Extension operations on a single virtual machine.
+    """
+    @property
+    def _ExtensionType(self) -> Type:
+        return VirtualMachineExtension
+
+    @property
+    def _begin_create_or_update(self) -> Callable[[str, str, str, VirtualMachineExtension], LROPoller[VirtualMachineExtension]]:
+        return self._compute_client.virtual_machine_extensions.begin_create_or_update
+
+    @property
+    def _get(self) -> VirtualMachineExtension:
+        return self._compute_client.virtual_machine_extensions.get
+
+    def delete(self) -> None:
+        log.info("Deleting %s", self._identifier)
+
+        execute_with_retry(lambda: self._compute_client.virtual_machine_extensions.begin_delete(
+            self._vm.resource_group,
+            self._vm.name,
+            self._resource_name
+        ).wait(timeout=_TIMEOUT))
+
+
+class VmssExtension(_VmExtensionBaseClass):
```

Comment:
This class has not been tested and is not used currently. Similar functionality was implemented in the prototype's azure_models.py, and I decided to port it as well, since for sure some tests will need it.